np.conj and transpose fails with different order

Using np.conj and transpose (x.T) fails if we do do conjugate first and transpose after. Testcase below:

This fails to compile:
```python
import numpy as np

#pythran export testconj(complex128[][])
def testconj(x):
    return np.conj(x).T
```

Error:
```
/tmp/tmpoxlnjb9q.cpp: In instantiation of ‘typename __pythran_testig::testconj::type<argument_type0>::result_type __pythran_testig::testconj::operator()(argument_type0&&) const [with argument_type0 = {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&; typename __pythran_testig::testconj::type<argument_type0>::result_type = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >]’:
/tmp/tmpoxlnjb9q.cpp:52:72:   required from here
/tmp/tmpoxlnjb9q.cpp:42:39: error: invalid use of incomplete type ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> >’}
   42 |     return pythonic::builtins::getattr(pythonic::types::attr::T{}, pythonic::numpy::functor::conj{}(x));
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/nested_container.hpp:6,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/tuple.hpp:9,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/builtins/bool_.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpoxlnjb9q.cpp:1:
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/numpy_traits.hpp:23:10: note: declaration of ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> >’}
   23 |   struct numpy_texpr;
      |          ^~~~~~~~~~~
/tmp/tmpoxlnjb9q.cpp: In instantiation of ‘typename __pythran_testig::testconj::type<argument_type0>::result_type __pythran_testig::testconj::operator()(argument_type0&&) const [with argument_type0 = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&; typename __pythran_testig::testconj::type<argument_type0>::result_type = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >]’:
/tmp/tmpoxlnjb9q.cpp:67:72:   required from here
/tmp/tmpoxlnjb9q.cpp:42:39: error: invalid use of incomplete type ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> >’}
   42 |     return pythonic::builtins::getattr(pythonic::types::attr::T{}, pythonic::numpy::functor::conj{}(x));
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/nested_container.hpp:6,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/tuple.hpp:9,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/builtins/bool_.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpoxlnjb9q.cpp:1:
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/numpy_traits.hpp:23:10: note: declaration of ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> >’}
   23 |   struct numpy_texpr;
      |          ^~~~~~~~~~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/ndarray.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/tuple.hpp:13,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:7,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpoxlnjb9q.cpp:1:
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp: In instantiation of ‘decltype ({anonymous}::pythonic::numpy::transpose(a)) {anonymous}::pythonic::builtins::getattr({anonymous}::pythonic::types::attr::T, const E&) [with E = {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&>; decltype ({anonymous}::pythonic::numpy::transpose(a)) = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> >]’:
/tmp/tmpoxlnjb9q.cpp:42:39:   required from ‘typename __pythran_testig::testconj::type<argument_type0>::result_type __pythran_testig::testconj::operator()(argument_type0&&) const [with argument_type0 = {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&; typename __pythran_testig::testconj::type<argument_type0>::result_type = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >]’
/tmp/tmpoxlnjb9q.cpp:52:72:   required from here
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp:828:8: error: return type ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> >’} is incomplete
  828 |   auto getattr(types::attr::T, E const &a) -> decltype(numpy::transpose(a))
      |        ^~~~~~~
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp:830:28: error: invalid use of incomplete type ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> >’}
  830 |     return numpy::transpose(a);
      |            ~~~~~~~~~~~~~~~~^~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/nested_container.hpp:6,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/tuple.hpp:9,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/builtins/bool_.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpoxlnjb9q.cpp:1:
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/numpy_traits.hpp:23:10: note: declaration of ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> >&> >’}
   23 |   struct numpy_texpr;
      |          ^~~~~~~~~~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/ndarray.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/tuple.hpp:13,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:7,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpoxlnjb9q.cpp:1:
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp: In instantiation of ‘decltype ({anonymous}::pythonic::numpy::transpose(a)) {anonymous}::pythonic::builtins::getattr({anonymous}::pythonic::types::attr::T, const E&) [with E = {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&>; decltype ({anonymous}::pythonic::numpy::transpose(a)) = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> >]’:
/tmp/tmpoxlnjb9q.cpp:42:39:   required from ‘typename __pythran_testig::testconj::type<argument_type0>::result_type __pythran_testig::testconj::operator()(argument_type0&&) const [with argument_type0 = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&; typename __pythran_testig::testconj::type<argument_type0>::result_type = {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >]’
/tmp/tmpoxlnjb9q.cpp:67:72:   required from here
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp:828:8: error: return type ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> >’} is incomplete
  828 |   auto getattr(types::attr::T, E const &a) -> decltype(numpy::transpose(a))
      |        ^~~~~~~
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp:830:28: error: invalid use of incomplete type ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> >’}
  830 |     return numpy::transpose(a);
      |            ~~~~~~~~~~~~~~~~^~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/nested_container.hpp:6,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/tuple.hpp:9,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/builtins/bool_.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpoxlnjb9q.cpp:1:
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/utils/numpy_traits.hpp:23:10: note: declaration of ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> > >::type’ {aka ‘struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::conjugate, {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::ndarray<std::complex<double>, {anonymous}::pythonic::types::pshape<long int, long int> > >&> >’}
   23 |   struct numpy_texpr;
      |          ^~~~~~~~~~~
WARNING: Compilation error, trying hard to find its origin...
WARNING: Nop, I'm going to flood you with C++ errors!
CRITICAL: Cover me Jack. Jack? Jaaaaack!!!!
```


switching the transpose and conjugate around works, so this is not a huge issue, but one can get easily caught out. This works:

```python
import numpy as np

#pythran export testconj(complex128[][])
def testconj(x):
    return np.conj(x.T)
```

--- a/pythran/pythonic/include/numpy/transpose.hpp
+++ b/pythran/pythonic/include/numpy/transpose.hpp
@@ -5,6 +5,7 @@
 #include "pythonic/include/utils/numpy_conversion.hpp"
 #include "pythonic/include/utils/nested_container.hpp"
 #include "pythonic/include/types/ndarray.hpp"
+#include "pythonic/include/types/numpy_expr.hpp"
 
 PYTHONIC_NS_BEGIN
 
@@ -38,6 +39,22 @@ namespace numpy
   template <class T, class pS, size_t M>
   types::ndarray<T, types::array<long, std::tuple_size<pS>::value>>
   transpose(types::ndarray<T, pS> const &a, types::array<long, M> const &t);
+
+  template <class Op, class... Args, size_t... Is>
+  auto _transpose(types::numpy_expr<Op, Args...> const &expr,
+                  utils::index_sequence<Is...>)
+      -> decltype(Op{}(transpose(std::get<Is>(expr.args)...)))
+  {
+    return Op{}(transpose(std::get<Is>(expr.args)...));
+  }
+
+  template <class Op, class... Args>
+  auto transpose(types::numpy_expr<Op, Args...> const &expr)
+      -> decltype(_transpose(expr,
+                             utils::make_index_sequence<sizeof...(Args)>()))
+  {
+    return _transpose(expr, utils::make_index_sequence<sizeof...(Args)>());
+  }
 
   template <class E>
   auto transpose(E const &expr) -> typename std::enable_if<

--- a/pythran/tests/test_numpy_func1.py
+++ b/pythran/tests/test_numpy_func1.py
@@ -190,6 +190,9 @@ class TestNumpyFunc1(TestEnv):
     def test_transpose_expr(self):
         self.run_test("def np_transpose_expr(a): return (a + a).transpose()", numpy.ones(24).reshape(2,3,4), np_transpose_expr=[NDArray[float,:,:,:]])
 
+    def test_transpose_expr2(self):
+        self.run_test("def np_transpose_expr2(a): import numpy as np; return np.conj(a).T", 1j * numpy.ones(6).reshape(2,3), np_transpose_expr2=[NDArray[complex,:,:]])
+
     def test_transpose2_(self):
         self.run_test("def np_transpose2_(a): return a.transpose((2,0,1))", numpy.arange(24).reshape(2,3,4), np_transpose2_=[NDArray[int,:,:,:]])
 

